stablehlo/dialect/AssemblyFormat.cpp

```diff
@@ -300,6 +301,38 @@ ParseResult parseDenseI64Array(OpAsmParser& parser,
   return success();
 }
 
+void printDynamicShape(AsmPrinter& p, llvm::ArrayRef<int64_t> shape) {
+  auto printIntOrQuestion = [&](int64_t value) {
+    if (ShapedType::isDynamic(value))
+      p << '?';
+    else
+      p << value;
+  };
+  p << " [";
+  llvm::interleaveComma(shape, p, printIntOrQuestion);
+  p << ']';
+}
+
+ParseResult parseDynamicShape(AsmParser& parser,
+                              FailureOr<SmallVector<int64_t>>& shapeResult) {
+  SmallVector<int64_t> shape;
+  auto parseElt = [&]() -> ParseResult {
```

Comment:
`parseIntOrQuestion`?

Reply:
`parseDimensionSizes` - Needs to match for custom declaration.